lsusb.c

```diff
@@ -320,6 +322,59 @@ static void dump_device(
 	free(serial);
 }
 
+
+/* Dump only serial number as text (mode==0) or hexadecimal (mode!=0) */
+static void dump_serial(libusb_device *dev, struct libusb_device_descriptor *desc, int mode)
+{
+	libusb_device_handle *udev;
+	if(!desc->iSerialNumber) {
+		if(mode)	// In hexadecimal mode print a hint
+			printf("NONE");
+		else
+			if (verblevel > 0)
+				fprintf(stderr, "Device without serial number descriptor\n");
+		return;
+	}
+	if (libusb_open(dev, &udev)) {
+		fprintf(stderr, "Couldn't open device, some information will be missing\n");
+		return;
+	} 
+
+	if(mode) {
+		// For hexadecimal output
+		unsigned char buf[255];	// Ahhrrg! Took me several hours, that size > 256 not works correct :-(
+		unsigned char *ptr=buf;	// Seems to me, that there is a &255 somewhere, but length parameter is an int !
+		int cnt=libusb_get_descriptor(udev,LIBUSB_DT_STRING,desc->iSerialNumber,buf,sizeof buf); 	
+		// Read raw descriptor instead of string. String-Verison:
+		// #define LANG_NEUTRAL 0x0000	// from https://docs.microsoft.com/en-us/windows/win32/intl/language-identifier-constants-and-strings
+  		// cnt=libusb_get_string_descriptor(udev,desc->iSerialNumber,LANG_NEUTRAL,buf,sizeof buf);
+		if(cnt<=2) {			// First two bytes are header: length and type=LIBUSB_DT_STRING.
+			printf("NULL");
+			fprintf(stderr, "Serial number descriptor has zero length or error\n");
+			return;
+		}
+		if((unsigned int)cnt>sizeof buf ) {
+			fprintf(stderr, "Error while reading serial number descriptor (size=%d)\n",cnt);
+			return;
+		}
+		ptr+=2; cnt-=2;		// Skip header
+		while(cnt--) {
+			// printf("%c",*ptr>31?*ptr:'?');
+			printf("%02x",*ptr);
+			ptr++;
+		}
+
+	} else {
+		// For text output
+		char *serial = get_dev_string(udev, desc->iSerialNumber);
```

Comment:
This is odd, the serial number should work to be retrieved this way as well.

I think your device is broken and does not follow the specification, as this is required to be a string and not "hex values", correct?  Have you asked if the device passes the USB certification process?

Reply:
I read about and agree that the ST.-Link/V2 device does abuse serial 
number field. String should be a UTF-16 LE text.
But in reality it contains e.g. 0x06 and 0x02 as a character code.
The higher bytes are always zero. 
Here my 2 devices with the modified lsusb:
5400ff006d00060072006600530053003000400002006700
4900ff007100060049007000530055004400460006008700
The device itself is working as it should with exception of the serial number.
I'm not the only one with this problem:
- https://github.com/texane/stlink/issues/417
- https://www.bergzand.net/programming-with-multiple-st-link-programmers.html